File: android/ac-charts/src/test/kotlin/ChartsTests.kt
package com.microsoft.adaptivecards.charts

import com.microsoft.adaptivecards.core.models.*
import kotlinx.serialization.json.Json
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class ChartsTests {
    
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }
    
    @Test
    fun `test DonutChart deserialization`() {
        val jsonString = """
        {
            "type": "DonutChart",
            "id": "chart1",
            "title": "Sales",
            "size": "medium",
            "showLegend": true,
            "innerRadiusRatio": 0.5,
            "data": [
                {"label": "A", "value": 10.0},
                {"label": "B", "value": 20.0}
            ]
        }
        """
        
        val chart = json.decodeFromString<DonutChart>(jsonString)
        
        assertEquals("chart1", chart.id)
        assertEquals("Sales", chart.title)
        assertEquals("medium", chart.size)
        assertEquals(true, chart.showLegend)
        assertEquals(0.5, chart.innerRadiusRatio)
        assertEquals(2, chart.data.size)
        assertEquals("A", chart.data[0].label)
        assertEquals(10.0, chart.data[0].value)
    }
    
    @Test
    fun `test BarChart deserialization`() {
        val jsonString = """
        {
            "type": "BarChart",
            "id": "chart2",
            "title": "Revenue",
            "orientation": "vertical",
            "showValues": true,
            "data": [
                {"label": "Jan", "value": 100.0},
                {"label": "Feb", "value": 200.0}
            ]
        }
        """
        
        val chart = json.decodeFromString<BarChart>(jsonString)
        
        assertEquals("chart2", chart.id)
        assertEquals("Revenue", chart.title)
        assertEquals("vertical", chart.orientation)
        assertEquals(true, chart.showValues)
        assertEquals(2, chart.data.size)
    }
    
    @Test
    fun `test LineChart deserialization`() {
        val jsonString = """
        {
            "type": "LineChart",
            "id": "chart3",
            "smooth": true,
            "showDataPoints": true,
            "data": [
                {"label": "Mon", "value": 50.0},
                {"label": "Tue", "value": 75.0}
            ]
        }
        """
        
        val chart = json.decodeFromString<LineChart>(jsonString)
        
        assertEquals("chart3", chart.id)
        assertEquals(true, chart.smooth)
        assertEquals(true, chart.showDataPoints)
        assertEquals(2, chart.data.size)
    }
    
    @Test
    fun `test PieChart deserialization`() {
        val jsonString = """
        {
            "type": "PieChart",
            "id": "chart4",
            "showPercentages": true,
            "data": [
                {"label": "A", "value": 30.0},
                {"label": "B", "value": 70.0}
            ]
        }
        """
        
        val chart = json.decodeFromString<PieChart>(jsonString)
        
        assertEquals("chart4", chart.id)
        assertEquals(true, chart.showPercentages)
        assertEquals(2, chart.data.size)
    }
    
    @Test
    fun `test chart with custom colors`() {
        val jsonString = """
        {
            "type": "DonutChart",
            "colors": ["#FF0000", "#00FF00", "#0000FF"],
            "data": [
                {"label": "Red", "value": 10.0},
                {"label": "Green", "value": 20.0},
                {"label": "Blue", "value": 30.0}
            ]
        }
        """
        
        val chart = json.decodeFromString<DonutChart>(jsonString)
        
        assertEquals(3, chart.colors?.size)
        assertEquals("#FF0000", chart.colors?.get(0))
    }
    
    @Test
    fun `test ChartColors default palette`() {
        assertEquals(8, ChartColors.defaultPalette.size)
    }
    
    @Test
    fun `test ChartSize enum`() {
        assertEquals(150, ChartSize.SMALL.heightDp)
        assertEquals(250, ChartSize.MEDIUM.heightDp)
        assertEquals(350, ChartSize.LARGE.heightDp)
        assertEquals(250, ChartSize.AUTO.heightDp)
        
        assertEquals(ChartSize.SMALL, ChartSize.from("small"))
        assertEquals(ChartSize.MEDIUM, ChartSize.from("medium"))
        assertEquals(ChartSize.LARGE, ChartSize.from("large"))
        assertEquals(ChartSize.AUTO, ChartSize.from("auto"))
        assertEquals(ChartSize.AUTO, ChartSize.from(null))
    }
}

File: ios/Tests/ACChartsTests/ChartsTests.swift
import XCTest
@testable import ACCore
@testable import ACCharts

final class ChartsTests: XCTestCase {

    func testDonutChartDecoding() throws {
        let json = """
        {
            "type": "DonutChart",
            "id": "chart1",
            "title": "Sales",
            "size": "medium",
            "showLegend": true,
            "innerRadiusRatio": 0.5,
            "data": [
                {"label": "A", "value": 10},
                {"label": "B", "value": 20}
            ]
        }
        """

        let data = json.data(using: .utf8)!
        let decoder = JSONDecoder()
        let element = try decoder.decode(CardElement.self, from: data)

        if case .donutChart(let chart) = element {
            XCTAssertEqual(chart.id, "chart1")
            XCTAssertEqual(chart.title, "Sales")
            XCTAssertEqual(chart.size, "medium")
            XCTAssertEqual(chart.showLegend, true)
            XCTAssertEqual(chart.innerRadiusRatio, 0.5)
            XCTAssertEqual(chart.data.count, 2)
            XCTAssertEqual(chart.data[0].label, "A")
            XCTAssertEqual(chart.data[0].value, 10)
        } else {
            XCTFail("Failed to decode DonutChart")
        }
    }

    func testBarChartDecoding() throws {
        let json = """
        {
            "type": "BarChart",
            "id": "chart2",
            "title": "Revenue",
            "orientation": "vertical",
            "showValues": true,
            "data": [
                {"label": "Jan", "value": 100},
                {"label": "Feb", "value": 200}
            ]
        }
        """

        let data = json.data(using: .utf8)!
        let decoder = JSONDecoder()
        let element = try decoder.decode(CardElement.self, from: data)

        if case .barChart(let chart) = element {
            XCTAssertEqual(chart.id, "chart2")
            XCTAssertEqual(chart.title, "Revenue")
            XCTAssertEqual(chart.orientation, "vertical")
            XCTAssertEqual(chart.showValues, true)
            XCTAssertEqual(chart.data.count, 2)
        } else {
            XCTFail("Failed to decode BarChart")
        }
    }

    func testLineChartDecoding() throws {
        let json = """
        {
            "type": "LineChart",
            "id": "chart3",
            "smooth": true,
            "showDataPoints": true,
            "data": [
                {"label": "Mon", "value": 50},
                {"label": "Tue", "value": 75}
            ]
        }
        """

        let data = json.data(using: .utf8)!
        let decoder = JSONDecoder()
        let element = try decoder.decode(CardElement.self, from: data)

        if case .lineChart(let chart) = element {
            XCTAssertEqual(chart.id, "chart3")
            XCTAssertEqual(chart.smooth, true)
            XCTAssertEqual(chart.showDataPoints, true)
            XCTAssertEqual(chart.data.count, 2)
        } else {
            XCTFail("Failed to decode LineChart")
        }
    }

    func testPieChartDecoding() throws {
        let json = """
        {
            "type": "PieChart",
            "id": "chart4",
            "showPercentages": true,
            "data": [
                {"label": "A", "value": 30},
                {"label": "B", "value": 70}
            ]
        }
        """

        let data = json.data(using: .utf8)!
        let decoder = JSONDecoder()
        let element = try decoder.decode(CardElement.self, from: data)

        if case .pieChart(let chart) = element {
            XCTAssertEqual(chart.id, "chart4")
            XCTAssertEqual(chart.showPercentages, true)
            XCTAssertEqual(chart.data.count, 2)
        } else {
            XCTFail("Failed to decode PieChart")
        }
    }

    func testChartWithCustomColors() throws {
        let json = """
        {
            "type": "DonutChart",
            "colors": ["#FF0000", "#00FF00", "#0000FF"],
            "data": [
                {"label": "Red", "value": 10},
                {"label": "Green", "value": 20},
                {"label": "Blue", "value": 30}
            ]
        }
        """

        let data = json.data(using: .utf8)!
        let decoder = JSONDecoder()
        let element = try decoder.decode(CardElement.self, from: data)

        if case .donutChart(let chart) = element {
            XCTAssertEqual(chart.colors?.count, 3)
            XCTAssertEqual(chart.colors?[0], "#FF0000")
        } else {
            XCTFail("Failed to decode chart with custom colors")
        }
    }
}
